dlt/common/plugins/plugins.py

```diff
@@ -0,0 +1,143 @@
+        return f(self, *args, **kwargs)
+
+    return _wrap  # type: ignore
+
+
+@configspec
+class PluginsContext(ContainerInjectableContext, SupportsCallbackPlugin):
+    def __init__(self, main: bool = True) -> None:
+        self._plugins: List[Plugin[BaseConfiguration]] = []
+        self._callback_plugins: List[CallbackPlugin[BaseConfiguration]] = []
+        self._initial_plugins: TPluginArg = []
+        self._main = main
+        self._queue: Optional[queue.Queue[Any]] = None
+
+    def _resolve_plugin(self, plugin: TSinglePluginArg) -> Plugin[BaseConfiguration]:
+        resolved_plugin: Plugin[BaseConfiguration] = None
+        if isinstance(plugin, str):
+            module_path, attr_name = plugin.rsplit(".", 1)
+            try:
+                module = import_module(module_path)
+            except ModuleNotFoundError as e:
+                raise UnknownPluginPathException(plugin) from e
+            try:
+                plugin = getattr(module, attr_name)
+            except AttributeError as e:
+                raise UnknownPluginPathException(str(plugin)) from e
+        if isinstance(plugin, type) and issubclass(plugin, Plugin):
+            resolved_plugin = plugin()
+        else:
+            raise TypeError(f"Plugin {plugin} is not a subclass of Plugin nor a plugin name string")
+        return resolved_plugin
+
+    # pickle support, send plugins config and queue to child process / threads
+    def __getstate__(self) -> Dict[str, Any]:
+        return {"plugins": self._initial_plugins, "queue": self._queue}
+
+    def __setstate__(self, d: Dict[str, Any]) -> None:
+        self.__init__(False)  # type: ignore[misc]
+        self.setup_plugins(d["plugins"])
+        self._queue = d["queue"]
+
+    def process_queue(self) -> None:
+        assert self._main
+        if not self._plugins:
+            return
+        try:
+            while True:
+                name, args, kwargs = self._queue.get_nowait()
+                getattr(self, (name + SAFE_SUFFIX))(*args, **kwargs)
+        except mp.queues.Empty:  # type: ignore
+            pass
+
+    def setup_plugins(self, plugins: TPluginArg) -> None:
+        if not plugins:
+            return
+        self._initial_plugins = plugins
+        # if not plugins:
+        if not isinstance(plugins, Iterable):
+            plugins = [plugins]
+        for p in plugins:
+            resolved_plugin = self._resolve_plugin(p)
+            self._plugins.append(resolved_plugin)
+            if isinstance(resolved_plugin, CallbackPlugin):
+                self._callback_plugins.append(resolved_plugin)
+
+        if self._plugins and self._main:
+            manager = mp.Manager()
+            self._queue = manager.Queue()
+
+    def get_plugin(self, plugin_name: str) -> Optional[Plugin[BaseConfiguration]]:
+        for p in self._plugins:
+            if p.NAME == plugin_name:
+                return p
+        return None
+
+    #
+    # Main Pipeline Callbacks
```

Comment:
Is it interesting to us to collect statistics about popular plugins in the future?